### PR TITLE
enhance e2e test and publish image

### PR DESCRIPTION
**Pull request by shivakunv**
**Old workflow :** 

        1.   Build local image  and push image to the repo 
        2.   e2e tests  
             a. upgrade kernel 
              b. fetch repo image on target with helm install to run driver tests

**New Workflow:**

      Build Local Image
          a. Build the image
          a. Save the built image as a .tar file using docker save.
          b. Upload the .tar file as an artifact (not to the repo).
      
      e2e Tests
          a. Upgrade the kernel on the target.
          b. Download the artifact (the saved .tar image file) from the CI pipeline
          c. Transfer the .tar image to the target using rsync.
          d. Import the .tar image on the target using the ctr command.
          e. Run helm install to execute driver tests.
      
      Publish or Push Image to Repo
          a. Download the artifact.
          b. Load the .tar image using docker load.
          c. Push the image to the repo using docker push.

### CODE REVIEW

File: .github/workflows/image.yaml [hunk 1]
```diff
@@ -16,17 +16,19 @@
...
-    types:
-      - opened
-      - synchronize
-    branches:
-      - main
-      - release-*
+  # SHIVA
+  # pull_request:
+  #   types:
+  #     - opened
```
Comment (shivakunv) [inline .github/workflows/image.yaml:22]: for testing

File: .github/workflows/precompiled.yaml [hunk 2]
```diff
@@ -17,7 +17,20 @@ name: Precompiled images
...
 on:
   schedule:
+    # SHIVA
     - cron: '00 09 * * *' # scheduled job
+  # pull_request:
+  #   types:
+  #     - opened
+  #     - synchronize
+  #   branches:
+  #     - main
```
Comment (shivakunv) [inline .github/workflows/precompiled.yaml:27]: test code

File: .github/workflows/precompiled.yaml [hunk 3]
```diff
@@ -31,22 +44,24 @@ jobs:
       - name: Read driver versions
         id: extract_driver_branch
         run: |
-          # get driver-branch
-          DRIVER_BRANCH=("535" "550")
+          # get driver_branch
+          # DRIVER_BRANCH=("535" "550")
+          DRIVER_BRANCH=("535")
```
Comment (shivakunv) [inline .github/workflows/precompiled.yaml:49]: test code

File: .github/workflows/precompiled.yaml [hunk 4]
```diff
@@ -31,22 +44,24 @@ jobs:
...
           driver_branch_json=$(printf '%s\n' "${DRIVER_BRANCH[@]}" | jq -R . | jq -cs .)
           echo "driver_branch=$driver_branch_json" >> $GITHUB_OUTPUT
 
           # get kernel flavors
-          KERNEL_FLAVORS=("aws" "azure" "generic" "nvidia" "oracle")
+          # KERNEL_FLAVORS=("aws" "azure" "generic" "nvidia" "oracle")
+          KERNEL_FLAVORS=("aws")
```
Comment (shivakunv) [inline .github/workflows/precompiled.yaml:55]: test code

File: .github/workflows/precompiled.yaml [hunk 5]
```diff
@@ -31,22 +44,24 @@ jobs:
...
           echo "kernel_flavors=$kernel_flavors_json" >> $GITHUB_OUTPUT
 
-  precompiled-image:
+  precompiled-build-image:
     needs: set-driver-version-matrix
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        driver-branch: ${{ fromJson(needs.set-driver-version-matrix.outputs.driver_branch) }}
+        driver_branch: ${{ fromJson(needs.set-driver-version-matrix.outputs.driver_branch) }}
```
Comment (shivakunv) [inline .github/workflows/precompiled.yaml:64]: using driver_branch in most of the places

File: .github/workflows/precompiled.yaml [hunk 6]
```diff
@@ -59,7 +74,7 @@ jobs:
           REPO_FULL_NAME="${{ github.repository }}"
           echo "LABEL_IMAGE_SOURCE=https://github.com/${REPO_FULL_NAME}" >> $GITHUB_ENV
 
-          GENERATE_ARTIFACTS="true"
+          GENERATE_ARTIFACTS="false"
```
Comment (shivakunv) [inline .github/workflows/precompiled.yaml:77]: do not upload

File: .github/workflows/precompiled.yaml [hunk 7]
```diff
@@ -220,19 +270,29 @@ jobs:
       - name: Precompiled e2e test gpu driver validation
         env:
           TEST_CASE: "./tests/cases/nvidia-driver.sh"
-          GPU_OPERATOR_OPTIONS: "--set driver.repository=${{ env.PRIVATE_REGISTRY }}/nvidia --set driver.usePrecompiled=true"
+          GPU_OPERATOR_OPTIONS: "--set driver.repository=${{ env.PRIVATE_REGISTRY }}/nvidia --set driver.usePrecompiled=true \
+            --set driver.imagePullPolicy=Never"
```
Comment (shivakunv) [inline .github/workflows/precompiled.yaml:274]: --set driver.imagePullPolicy=Never   : use local image uploaded with ctr command on target

File: .github/workflows/precompiled.yaml [hunk 8]
```diff
@@ -220,19 +270,29 @@ jobs:
...
           rc=0
           # for precompiled driver we are setting driver branch as driver version
-          driver_versions_json='${{ needs.set-driver-version-matrix.outputs.driver_branch }}'
-          driver_versions=$(echo "$driver_versions_json" | jq -r '.[]')
-          for DRIVER_VERSION in $driver_versions; do
+          for DRIVER_VERSION in $driver_branch; do
             echo "Running e2e for DRIVER_VERSION=$DRIVER_VERSION"
+            # image="driver-images-${DRIVER_VERSION}-${KERNEL_VERSION}-${DIST}"
+            # SHIVA remove hardcode value one is for testing
+            image="driver-images-${DRIVER_VERSION}-5.15.0-1069-aws-ubuntu22.04"
```
Comment (shivakunv) [inline .github/workflows/precompiled.yaml:283]: test code

File: .github/workflows/precompiled.yaml [hunk 9]
```diff
@@ -220,19 +270,29 @@ jobs:
...
             status=0
-            OPERATOR_OPTIONS="${GPU_OPERATOR_OPTIONS} --set driver.version=${DRIVER_VERSION}"
+            TEST_CASE_ARGS="${GPU_OPERATOR_OPTIONS} --set driver.version=${DRIVER_VERSION}"
             # add escape character for space
-            OPERATOR_OPTIONS=$(printf '%q ' "$OPERATOR_OPTIONS")
-            ./tests/ci-run-e2e.sh "${TEST_CASE}" "${OPERATOR_OPTIONS}" || status=$?
+            TEST_CASE_ARGS=$(printf '%q ' "$TEST_CASE_ARGS")
+            # SHIVA
+            # IMAGE_PATH="./tests/driver-images-${DRIVER_VERSION}-${KERNEL_VERSION}-${DIST}.tar"
+            IMAGE_PATH="./tests/driver-images-${DRIVER_VERSION}-5.15.0-1069-aws-ubuntu22.04.tar"
```
Comment (shivakunv) [inline .github/workflows/precompiled.yaml:294]: test code

File: .github/workflows/precompiled.yaml [hunk 10]
```diff
@@ -248,3 +308,47 @@ jobs:
...
+        run: |
+          echo "PRIVATE_REGISTRY=ghcr.io" >> $GITHUB_ENV
+          echo "DIST=ubuntu22.04" >> $GITHUB_ENV
+ 
+      - name: Download built image artifact
+        uses: actions/download-artifact@v4
+        with:
+          # SHIVA
+          #  name: driver-images-${{ matrix.driver_branch }}-${{ matrix.kernel_version }}-${{ env.DIST }}
+           name: driver-images-${{ matrix.driver_branch }}-5.15.0-1069-aws-ubuntu22.04
```
Comment (shivakunv) [inline .github/workflows/precompiled.yaml:342]: test code

File: tests/scripts/findkernelversion.sh [hunk 11]
```diff
@@ -30,3 +30,5 @@ if [[ $status -eq 0 ]]; then
 else
     export should_continue=true
 fi
+# SHIVA
+export should_continue=true
```
Comment (shivakunv) [inline tests/scripts/findkernelversion.sh:34]: test code

File: tests/scripts/findkernelversion.sh [hunk 12]
```diff
@@ -22,11 +22,11 @@ export $(grep -oP 'KERNEL_VERSION=[^ ]+' ./kernel_version.txt)
...
-echo "regctl tag ls  nvcr.io/nvidia/driver | grep "^${DRIVER_BRANCH}-${KERNEL_VERSION}-${DIST}$""
-
 regctl tag ls  nvcr.io/nvidia/driver | grep "^${DRIVER_BRANCH}-${KERNEL_VERSION}-${DIST}$" || status=$?
 if [[ $status -eq 0 ]]; then
     export should_continue=false
 else
     export should_continue=true
 fi
+# SHIVA
+export should_continue=true
```
Comment (shivakunv) [inline tests/scripts/findkernelversion.sh:32]: test code

File: .github/workflows/precompiled.yaml [hunk 13]
```diff
@@ -93,19 +100,44 @@ jobs:
...
+          make DRIVER_VERSIONS=${DRIVER_VERSIONS} DRIVER_BRANCH=${{ matrix.driver_branch }} build-${DIST}-${DRIVER_VERSION}
+
+      - name: Save build image as a tar
+        env:
+          DIST: ${{ matrix.dist }}
+          PRIVATE_REGISTRY: "ghcr.io"
+          COMMIT_SHORT_SHA: ${COMMIT_SHORT_SHA}
+        run: |
+          source kernel_version.txt
+          docker images "${PRIVATE_REGISTRY}/nvidia/driver:${{ matrix.driver_branch }}-${KERNEL_VERSION}-${DIST}"
```
Comment (cdesiniotis) [inline .github/workflows/precompiled.yaml:115]: Question -- is this line required?

File: .github/workflows/precompiled.yaml [hunk 14]
```diff
@@ -120,42 +152,28 @@ jobs:
...
+          DRIVER_BRANCHES=($(echo "$driver_branch_json" | jq -r '.[]'))
 
+          source ./tests/scripts/ci-precompiled-helpers.sh
+          KERNEL_VERSIONS=($(get_kernel_versions_to_test $BASE_TARGET KERNEL_FLAVORS[@] DRIVER_BRANCHES[@] $DIST))
+          if [ -z "$KERNEL_VERSIONS" ]; then
+            # no new kernel release
+            echo "Skipping e2e tests"
+            exit 0
+          fi
+          echo "matrix_values_not_empty=1" >> $GITHUB_OUTPUT
```
Comment (cdesiniotis) [inline .github/workflows/precompiled.yaml:171]: This command is a duplicate of line 176. We can remove this line.

File: .github/workflows/precompiled.yaml (hunk 13 above)
Comment (shivakunv) [reply .github/workflows/precompiled.yaml:115]: to build local docker image(   do not push with flag GENERATE_ARTIFACTS="false") for docker save command,  which uploads .tar image as artifacts

File: .github/workflows/precompiled.yaml (hunk 14 above)
Comment (shivakunv) [reply .github/workflows/precompiled.yaml]: removed . Thanks

File: .github/workflows/precompiled.yaml (hunk 13 above)
Comment (cdesiniotis) [reply .github/workflows/precompiled.yaml:115]: But why is this needed? Doesn't `docker images` just list the image? Is this to verify the tag exists before attempting to save? If it doesn't exist I would expect the save command to return an error.

Comment (shivakunv) [reply .github/workflows/precompiled.yaml:115]: docker images will list the images, but we need a locally built image for the current code for e2etests.
Also, if we want to e2etest an earlier release from the repo, we need to run docker pull to get the old image. 

**Old workflow :** 

1.   Build local image (current code build) and push image to the repo 
2.   e2e tests  
     a. upgrade kernel 
      b. fetch repo image on target with helm install to run driver tests

**New Workflow:**

**Build Local Image (Current Code Build)**
    a. Build the image
    a. Save the built image as a .tar file using docker save.
    b. Upload the .tar file as an artifact (not to the repo).

**e2e Tests**
    a. Upgrade the kernel on the target.
    b. Download the artifact (the saved .tar image file) from the CI pipeline
    c. Transfer the .tar image to the target using rsync.
    d. Import the .tar image on the target using the ctr command.
    e. Run helm install to execute driver tests.

**Publish or Push Image to Repo**
    a. Download the artifact.
    b. Load the .tar image using docker load.
    c. Push the image to the repo using docker push.

Comment (shivakunv) [reply .github/workflows/precompiled.yaml:115]: Thanks .   removed 

The docker images command is not required (it was previously added for debugging purposes).
I was confused about the command docker images and the make command

**Review by cdesiniotis — APPROVED** [review]

**Pull request merged**